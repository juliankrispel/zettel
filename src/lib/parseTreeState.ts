import { FlatState, TreeState, Block } from './types'
import id from './id'

const getNode = (state: TreeState, path: number[]): Block  => {
  return path.reduce((acc: any, val) => {
    return acc.nodes[val]
  }, state)
}


const getNodes = (state: TreeState, path: number[]): Block[]  => {
  if (path.length === 0) {
    return state.nodes
  } else {
    return path.reduce((acc, val) => {
      return acc[val].nodes || acc
    }, state.nodes)
  }
}

const parseTreeState = (flat: FlatState): TreeState => {
  const state: TreeState = {
    nodes: [],
    meta: {},
  }

  let path: number[] = []

  flat.value.forEach((char, index) => {
    // if we have a start block character,
    // create a new block and add it
    if (char.char === '[') {
      const nodes = getNodes(state, path)

      nodes.push({
        value: [],
        nodes: [],
        key: id(),
      })

      path.push(nodes.length - 1)
    } else if (char.char === ']') {
      path.pop()
    } else {
      const node = getNode(state, path) || state
      node.value.push({
        char: char.char,
        styles: []
      })
    }
  })

  return state
}

export default parseTreeState