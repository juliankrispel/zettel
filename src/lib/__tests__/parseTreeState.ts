import parseTreeState from '../parseTreeState'

jest.mock('../id', () => {
  var i = 0
  return () => `key-${i++}`
})

describe('parseTreeState', () => {
  test('parses simple flat structure', () => {
    const flatState = {
      value: [
        { char: '[', metaKeys: [] },
        { char: 'H', metaKeys: [] },
        { char: 'e', metaKeys: [] },
        { char: 'l', metaKeys: [] },
        { char: 'l', metaKeys: [] },
        { char: 'o', metaKeys: [] },
        { char: ']', metaKeys: [] },
      ],
      meta: {}
    }

    expect(parseTreeState(flatState)).toMatchSnapshot()
  })

  test('parses nested structure', () => {
    const flatState = {
      value: [
        { char: '[', metaKeys: [] },
        { char: '1', metaKeys: [] },
        { char: '[', metaKeys: [] },
        { char: '2', metaKeys: [] },
        { char: '[', metaKeys: [] },
        { char: '3', metaKeys: [] },
        { char: ']', metaKeys: [] },
        { char: '[', metaKeys: [] },
        { char: '4', metaKeys: [] },
        { char: ']', metaKeys: [] },
        { char: ']', metaKeys: [] },
        { char: ']', metaKeys: [] },
      ],
      meta: {}
    }

    expect(parseTreeState(flatState)).toMatchSnapshot()
  })

  test('parses invalid structure', () => {
    const flatState = {
      value: [
        { char: '[', metaKeys: [] },
        { char: 'H', metaKeys: [] },
        { char: '[', metaKeys: [] },
        { char: 'l', metaKeys: [] },
        { char: 'l', metaKeys: [] },
        { char: 'o', metaKeys: [] },
        { char: ']', metaKeys: [] },
      ],
      meta: {}
    }

    console.log(JSON.stringify(parseTreeState(flatState), null, 2))
  })
})