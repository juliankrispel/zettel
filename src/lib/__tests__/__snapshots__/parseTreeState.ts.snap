// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseTreeState nested works 1`] = `
Object {
  "meta": Object {},
  "nodes": Array [
    Object {
      "_flatOffset": 0,
      "key": "key-1",
      "nodes": Array [
        Object {
          "_flatOffset": 2,
          "key": "key-2",
          "nodes": Array [
            Object {
              "_flatOffset": 4,
              "key": "key-3",
              "nodes": Array [],
              "value": Array [
                Object {
                  "char": "3",
                  "metaKeys": Array [],
                },
              ],
            },
            Object {
              "_flatOffset": 7,
              "key": "key-4",
              "nodes": Array [],
              "value": Array [
                Object {
                  "char": "4",
                  "metaKeys": Array [],
                },
              ],
            },
          ],
          "value": Array [
            Object {
              "char": "2",
              "metaKeys": Array [],
            },
          ],
        },
      ],
      "value": Array [
        Object {
          "char": "1",
          "metaKeys": Array [],
        },
      ],
    },
  ],
}
`;

exports[`parseTreeState parses  1`] = `
Object {
  "meta": Object {},
  "nodes": Array [
    Object {
      "_flatOffset": 0,
      "key": "key-0",
      "nodes": Array [],
      "value": Array [
        Object {
          "char": "H",
          "metaKeys": Array [],
        },
        Object {
          "char": "e",
          "metaKeys": Array [],
        },
        Object {
          "char": "l",
          "metaKeys": Array [],
        },
        Object {
          "char": "l",
          "metaKeys": Array [],
        },
        Object {
          "char": "o",
          "metaKeys": Array [],
        },
      ],
    },
  ],
}
`;

exports[`parseTreeState parses nested structure 1`] = `
Object {
  "meta": Object {},
  "nodes": Array [
    Object {
      "_flatOffset": 0,
      "key": "key-1",
      "nodes": Array [
        Object {
          "_flatOffset": 2,
          "key": "key-2",
          "nodes": Array [
            Object {
              "_flatOffset": 4,
              "key": "key-3",
              "nodes": Array [],
              "value": Array [
                Object {
                  "char": "3",
                  "metaKeys": Array [],
                },
              ],
            },
            Object {
              "_flatOffset": 7,
              "key": "key-4",
              "nodes": Array [],
              "value": Array [
                Object {
                  "char": "4",
                  "metaKeys": Array [],
                },
              ],
            },
          ],
          "value": Array [
            Object {
              "char": "2",
              "metaKeys": Array [],
            },
          ],
        },
      ],
      "value": Array [
        Object {
          "char": "1",
          "metaKeys": Array [],
        },
      ],
    },
  ],
}
`;

exports[`parseTreeState parses simple  1`] = `
Object {
  "meta": Object {},
  "nodes": Array [
    Object {
      "_flatOffset": 0,
      "key": "key-0",
      "nodes": Array [],
      "value": Array [
        Object {
          "char": "H",
          "metaKeys": Array [],
        },
        Object {
          "char": "e",
          "metaKeys": Array [],
        },
        Object {
          "char": "l",
          "metaKeys": Array [],
        },
        Object {
          "char": "l",
          "metaKeys": Array [],
        },
        Object {
          "char": "o",
          "metaKeys": Array [],
        },
      ],
    },
  ],
}
`;

exports[`parseTreeState parses simple flat structure 1`] = `
Object {
  "meta": Object {},
  "nodes": Array [
    Object {
      "_flatOffset": 0,
      "key": "key-0",
      "nodes": Array [],
      "value": Array [
        Object {
          "char": "H",
          "metaKeys": Array [],
        },
        Object {
          "char": "e",
          "metaKeys": Array [],
        },
        Object {
          "char": "l",
          "metaKeys": Array [],
        },
        Object {
          "char": "l",
          "metaKeys": Array [],
        },
        Object {
          "char": "o",
          "metaKeys": Array [],
        },
      ],
    },
  ],
}
`;

exports[`parseTreeState parses simple flat tree 1`] = `
Object {
  "meta": Object {},
  "nodes": Array [
    Object {
      "_flatOffset": 0,
      "key": "key-0",
      "nodes": Array [],
      "value": Array [
        Object {
          "char": "H",
          "metaKeys": Array [],
        },
        Object {
          "char": "e",
          "metaKeys": Array [],
        },
        Object {
          "char": "l",
          "metaKeys": Array [],
        },
        Object {
          "char": "l",
          "metaKeys": Array [],
        },
        Object {
          "char": "o",
          "metaKeys": Array [],
        },
      ],
    },
  ],
}
`;

exports[`parseTreeState works 1`] = `
Object {
  "meta": Object {},
  "nodes": Array [
    Object {
      "_flatOffset": 0,
      "key": "key-0",
      "nodes": Array [],
      "value": Array [
        Object {
          "char": "H",
          "metaKeys": Array [],
        },
        Object {
          "char": "e",
          "metaKeys": Array [],
        },
        Object {
          "char": "l",
          "metaKeys": Array [],
        },
        Object {
          "char": "l",
          "metaKeys": Array [],
        },
        Object {
          "char": "o",
          "metaKeys": Array [],
        },
      ],
    },
  ],
}
`;
